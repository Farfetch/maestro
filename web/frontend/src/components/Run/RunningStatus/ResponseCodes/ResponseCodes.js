import { Badge, Collapse, Table, Typography } from "antd";
import React, { useEffect, useState } from "react";

import { fetchMetrics } from "../../../../lib/api/endpoints/runMetric";
import PageSpinner from "../../../layout/PageSpinner";

const { Panel } = Collapse;
const { Text } = Typography;

const ResponseCodes = ({ runId }) => {
  const [errorCodes, setErrorCodes] = useState([]);
  const [isLoading, setIsLoading] = useState(false);

  const updateUrlMetrics = async (runIdToFetch) => {
    setIsLoading(true);

    const metricsRes = await fetchMetrics(runIdToFetch, 0, true);

    const filterMetrics = (metricsArray) => {
      const responseCodeGroups = {};

      metricsArray.forEach((metric) => {
        metric.responses.forEach((response) => {
          const { responseCode } = response;

          if (!responseCodeGroups[responseCode]) {
            responseCodeGroups[responseCode] = [];
          }

          responseCodeGroups[responseCode].push(metric);
        });
      });

      return responseCodeGroups;
    };

    const formattedUrlMetrics = metricsRes.map(
      ({ totalCount, successCount, label, responses }) => {
        const errorsCount = totalCount - successCount;

        return {
          key: label,
          errorsCount,
          responses
        };
      }
    );

    const errorMetrics = filterMetrics(formattedUrlMetrics);

    setErrorCodes(errorMetrics);

    setIsLoading(false);
  };

  useEffect(() => {
    updateUrlMetrics(runId);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [runId]);

  const columns = [
    {
      title: "Label",
      dataIndex: "label",
      key: "label",
      ellipsis: true,
      sorter: (a, b) => a.label.localeCompare(b.label)
    },
    {
      title: "Response Messages",
      dataIndex: "responses",
      key: "responses",
      render: (responses) => (
        <div>
          {responses.map((response, idx) => (
            <Text key={`${response.responseCode}-${idx}`}>
              {response.messages.map((message, msgIdx) => (
                <React.Fragment
                  key={`${response.responseCode}-${idx}-${msgIdx}`}
                >
                  {message}
                  {msgIdx !== response.messages.length - 1 && <br />}
                </React.Fragment>
              ))}
              <br />
            </Text>
          ))}
        </div>
      )
    },
    {
      title: "Errors Count",
      dataIndex: "errorsCount",
      key: "errorsCount",
      width: 150,
      defaultSortOrder: "descend",
      sorter: (a, b) => a.errorsCount - b.errorsCount
    }
  ];

  return (
    <>
      <div>
        <h3>{`Error Codes ( Filter applied: < 100 and > 400 )`}</h3>
      </div>
      {isLoading ? (
        <PageSpinner />
      ) : (
        <Collapse accordion>
          {Object.entries(errorCodes).map(([responseCode, metrics]) => {
            const numericResponseCode = parseInt(responseCode, 10);

            if (
              Number.isNaN(numericResponseCode) ||
              numericResponseCode < 100 ||
              numericResponseCode > 399 ||
              typeof responseCode !== "string"
            ) {
              // Calculate the total errors count for this group
              const totalErrorsCount = metrics.reduce(
                (acc, metric) => acc + metric.errorsCount,
                0
              );

              return (
                <Panel
                  header={
                    <span>
                      {responseCode}{" "}
                      <Badge count={totalErrorsCount} overflowCount={99999} />
                    </span>
                  }
                  key={responseCode}
                >
                  <div key={`table-${responseCode}`}>
                    <Table
                      dataSource={metrics.map((metric, index) => ({
                        key: `${metric.key}-${index}`,
                        label: metric.key,
                        errorsCount: metric.errorsCount,
                        responses: metric.responses.filter(
                          (response) =>
                            response.responseCode.toString() === responseCode
                        )
                      }))}
                      columns={columns}
                      pagination={false}
                    />
                  </div>
                </Panel>
              );
            }

            return null;
          })}
        </Collapse>
      )}
    </>
  );
};

export default ResponseCodes;
