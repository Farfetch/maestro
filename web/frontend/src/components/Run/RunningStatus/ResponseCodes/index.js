import ResponseCodes from "./ResponseCodes";

export default ResponseCodes;
