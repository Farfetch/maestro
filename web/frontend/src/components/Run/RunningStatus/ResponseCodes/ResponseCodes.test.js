import { render } from "@testing-library/react";
import React from "react";

import ResponseCodes from "./ResponseCodes";

describe("ResponseCodes", () => {
  const mockErrorCodes = {
    500: [{ key: "error1", errorsCount: 5, responses: [] }],
    501: [{ key: "error2", errorsCount: 3, responses: [] }]
  };

  test("should render ResponseCodes component", () => {
    const rendered = render(<ResponseCodes errorCodes={mockErrorCodes} />);
    const component = rendered.container;
    expect(component.outerHTML).toMatchSnapshot();
  });
});
