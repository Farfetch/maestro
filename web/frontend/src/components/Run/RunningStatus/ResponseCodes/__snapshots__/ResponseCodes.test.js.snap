// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponseCodes should render ResponseCodes component 1`] = `"<div><div><h3>Error Codes ( Filter applied: &lt; 100 and &gt; 400 )</h3></div><div class=\\"ant-collapse ant-collapse-icon-position-left\\" role=\\"tablist\\"><div class=\\"ant-collapse-item\\"><div class=\\"ant-collapse-header\\" role=\\"tab\\" tabindex=\\"0\\" aria-expanded=\\"false\\"><div><span role=\\"img\\" aria-label=\\"right\\" class=\\"anticon anticon-right ant-collapse-arrow\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"right\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z\\"></path></svg></span></div><span>500 <span class=\\"ant-badge ant-badge-not-a-wrapper\\"><sup data-show=\\"true\\" class=\\"ant-scroll-number ant-badge-count\\" title=\\"5\\"><span class=\\"ant-scroll-number-only\\" style=\\"transition: none;\\"><span class=\\"ant-scroll-number-only-unit current\\">5</span></span></sup></span></span></div></div><div class=\\"ant-collapse-item\\"><div class=\\"ant-collapse-header\\" role=\\"tab\\" tabindex=\\"0\\" aria-expanded=\\"false\\"><div><span role=\\"img\\" aria-label=\\"right\\" class=\\"anticon anticon-right ant-collapse-arrow\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"right\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z\\"></path></svg></span></div><span>501 <span class=\\"ant-badge ant-badge-not-a-wrapper\\"><sup data-show=\\"true\\" class=\\"ant-scroll-number ant-badge-count\\" title=\\"3\\"><span class=\\"ant-scroll-number-only\\" style=\\"transition: none;\\"><span class=\\"ant-scroll-number-only-unit current\\">3</span></span></sup></span></span></div></div></div></div>"`;
